File: app/src/test/java/com/coinninja/coinkeeper/receiver/ApplicationStartedReceiverTest.java
package com.coinninja.coinkeeper.receiver;

import android.content.Intent;

import com.coinninja.coinkeeper.TestCoinKeeperApplication;
import com.coinninja.coinkeeper.service.PushNotificationEndpointRegistrationService;

import org.junit.After;
import org.junit.Before;
import org.junit.Test;
import org.junit.runner.RunWith;
import org.mockito.ArgumentCaptor;
import org.mockito.MockitoAnnotations;
import org.robolectric.RobolectricTestRunner;
import org.robolectric.RuntimeEnvironment;
import org.robolectric.annotation.Config;

import static org.hamcrest.MatcherAssert.assertThat;
import static org.hamcrest.Matchers.equalTo;
import static org.mockito.Matchers.eq;
import static org.mockito.Mockito.verify;
import static org.mockito.Mockito.verifyZeroInteractions;
import static org.mockito.Mockito.when;

@RunWith(RobolectricTestRunner.class)
@Config(application = TestCoinKeeperApplication.class)
public class ApplicationStartedReceiverTest {

    TestCoinKeeperApplication application;

    private ApplicationStartedReceiver appStartedReceiver;

    @Before
    public void setUp() {
        MockitoAnnotations.initMocks(this);
        application = (TestCoinKeeperApplication) RuntimeEnvironment.application;
        appStartedReceiver = new ApplicationStartedReceiver();
    }

    @After
    public void tearDown() {
        application = null;
        appStartedReceiver = null;
    }

    @Test
    public void schedules_push_notification_job_intent_services() {
        when(application.cnWalletManager.hasWallet()).thenReturn(true);
        ArgumentCaptor<Intent> captor = ArgumentCaptor.forClass(Intent.class);

        appStartedReceiver.onReceive(application, null);

        verify(application.jobServiceScheduler).enqueueWork(eq(application),
                eq(PushNotificationEndpointRegistrationService.class),
                eq(100), captor.capture());
        Intent intent = captor.getValue();
        assertThat(intent.getComponent().getClassName(), equalTo(PushNotificationEndpointRegistrationService.class.getName()));
    }

    @Test
    public void skip_when_missing_wallet() {
        when(application.cnWalletManager.hasWallet()).thenReturn(false);

        appStartedReceiver.onReceive(application, null);

        verifyZeroInteractions(application.jobServiceScheduler);
        verifyZeroInteractions(application.syncWalletManager);

    }

}

File: app/src/test/java/com/coinninja/coinkeeper/receiver/ApplicationStartedReceiverTest.kt
package com.coinninja.coinkeeper.receiver

import android.content.Intent
import androidx.test.core.app.ApplicationProvider
import androidx.test.ext.junit.runners.AndroidJUnit4
import com.coinninja.coinkeeper.TestCoinKeeperApplication
import com.coinninja.coinkeeper.service.PushNotificationEndpointRegistrationService
import com.coinninja.coinkeeper.util.analytics.Analytics
import com.nhaarman.mockitokotlin2.*
import org.hamcrest.MatcherAssert.assertThat
import org.hamcrest.Matchers.equalTo
import org.junit.Test
import org.junit.runner.RunWith

@RunWith(AndroidJUnit4::class)
class ApplicationStartedReceiverTest {

    private val application: TestCoinKeeperApplication get() = ApplicationProvider.getApplicationContext()

    private fun createReceiver(): ApplicationStartedReceiver {
        val receiver = ApplicationStartedReceiver()
        val application1 = application
        application1.jobServiceScheduler = mock()
        application1.syncWalletManager = mock()
        application1.cnWalletManager = mock()
        application1.analytics = mock()
        return receiver
    }

    @Test
    fun schedules_push_notification_job_intent_services() {
        val receiver = createReceiver()
        val app = application
        whenever(app.cnWalletManager.hasWallet).thenReturn(true)
        val captor = argumentCaptor<Intent>()

        receiver.onReceive(app, Intent())

        verify(app.jobServiceScheduler).enqueueWork(eq(app),
                eq(PushNotificationEndpointRegistrationService::class.java),
                eq(100), captor.capture())

        val intent = captor.firstValue
        assertThat(intent.component!!.javaClass.name, equalTo(PushNotificationEndpointRegistrationService::class.java.name))
    }

    @Test
    fun skip_when_missing_wallet() {
        val receiver = createReceiver()
        val app = application
        whenever(app.cnWalletManager.hasWallet).thenReturn(false)

        receiver.onReceive(app, Intent())

        verifyZeroInteractions(app.jobServiceScheduler)
        verifyZeroInteractions(app.syncWalletManager)
    }

    @Test
    fun sets_platform_to_android() {
        val receiver = createReceiver()
        val app = application
        whenever(app.cnWalletManager.hasWallet).thenReturn(false)
        verify(app.analytics).setUserProperty(Analytics.PROPERTY_PLATFORM, "Android")
    }
}
